### ISSUE TITLE
compiled_frame memory leak when multiple threads call a function that throws an exception

### ISSUE THREAD

**Issue by Kenrick0** [Nuitka/Nuitka#871]
Example that demonstrates the issue (this leaks memory at ~1MB/s):
```
import threading
import time

def func_a():
    time.sleep(0.001)
    raise Exception()

def thread_main():
    while True:
        try:
            func_a()
        except Exception:
            pass

a = threading.Thread(target=thread_main)
a.start()

thread_main()
```

Reproduced the issue on the versions below. I'm not aware of any version that doesn't have this bug.
Nuitka installed from pip, python from apt. Tried both with and without virtualenv, same result.
It does not leak running as a .py source file under python or python3.

```
0.6.10rc5 (develop branch)
Python: 2.7.17 (default, Sep 30 2020, 13:38:04), OS: Linux, Arch: x86_64

0.6.9.4
Python: 2.7.13 (default, Sep 26 2018, 18:42:22), OS: Linux, Arch: armv7l

0.6.9.4
Python: 3.5.3 (default, Sep 27 2018, 17:25:39), OS: Linux, Arch: armv7l

0.6.1
Python: 2.7.9 (default, Sep 16 2019, 18:29:54), OS: Linux, Arch: armv7l
```

It seems that it only occurs when two threads are in the same function at the same time and at least one raises an exception that causes it to exit the function. If they both raise an exception it appears to leak ~2x as fast. I have tracked this into the generated `template_frame_guard_full_exception_handler` code from `nuitka/codegen/templates/CodeTemplateFrames.py` where it doesn't free the allocated frame unless it equals the cache frame.

The solution may be as simple as moving the line `%(frame_cache_identifier)s = NULL;` inside the if condition, I made this change and it appeared to fix the leak for the above sample but I'm not sure if there are side effects of this. I also note that there is almost identical code further down that file for `template_frame_guard_generator_exception_handler` that may have the same bug (I have not tested this).

**Comment by kayhayen** [comment]
Wow, very nice analysis and good catch.

**Comment by kayhayen** [comment]
I just included it in the latest pre-release, as there is no way it's wrong, and on the other hand explains your observation perfectly fine, although I have to admit, I didn't look into it at run time, so please report back if it really helps you.

**Comment by Kenrick0** [comment]
Thanks for including the change, it seems like it works to me. I tested the sample above on Python 2.7.13 and Python 3.6.9 and didn't see either leaking. I also recompiled the original program and it also doesn't appear to be leaking.

I ran into a new issue attempting to run the develop branch 0.6.10rc6 on a Linux/Ubuntu x86_64 test machine that doesn't have ccache.
```
python3 -m nuitka test.py
Nuitka:INFO: Starting Python compilation.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating C source code for backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C level backend compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: gcc (gcc).
Nuitka will make use of ccache to speed up repeated compilation.

Is it OK to download and put it in "/home/kenrick/.local/share/Nuitka/ccache/v3.7.12".

No installer needed, cached, one time question.

Proceed and download? [Yes]/No 
No
TypeError: stat: path should be string, bytes, os.PathLike or integer, not NoneType:
  File "/home/kenrick/nuitka_tests/VENV3/lib/python3.6/site-packages/nuitka/build/SingleExe.scons", line 1744:
    assume_yes_for_downloads=assume_yes_for_downloads,
  File "/home/kenrick/nuitka_tests/VENV3/lib/python3.6/site-packages/nuitka/build/SconsCaching.py", line 203:
    assume_yes_for_downloads=assume_yes_for_downloads,
  File "/home/kenrick/nuitka_tests/VENV3/lib/python3.6/site-packages/nuitka/build/SconsCaching.py", line 117:
    assume_yes_for_downloads=assume_yes_for_downloads,
  File "/home/kenrick/nuitka_tests/VENV3/lib/python3.6/site-packages/nuitka/utils/Download.py", line 119:
    addFileExecutablePermission(exe_path)
  File "/home/kenrick/nuitka_tests/VENV3/lib/python3.6/site-packages/nuitka/utils/FileOperations.py", line 359:
    old_stat = os.stat(filename)
```

If I say 'yes' it responds with the following, can't imagine that's going to work on linux.
```
Nuitka:INFO: Downloading 'https://github.com/ccache/ccache/releases/download/v3.7.12/ccache-3.7.12-windows-32.zip'
Nuitka:INFO: Extracting to '/home/kenrick/.local/share/Nuitka/ccache/v3.7.12/ccache.exe'
sh: 1: /home/kenrick/.local/share/Nuitka/ccache/v3.7.12/ccache.exe: Exec format error
scons: *** [__constants.o] Error 2
```

I made the following change in `utils/Download.py` on line 112 so I can say 'no' without it failing. (alternatively I could `apt install ccache`)
```
    # Check success here.
    if os.path.isfile(exe_path):
        addFileExecutablePermission(exe_path)
    else:
        if reject:
            sys.exit(reject)

        exe_path = None
```

**Comment by kayhayen** [comment]
Hah, no ccache on Linux isn't well tested apparently. :)

**Comment by kayhayen** [comment]
Released as part of 0.6.10
